staging/src/k8s.io/apiserver/pkg/cel/environment/environment.go

```diff
@@ -0,0 +1,249 @@
+	//
+	// If an added cel.Variable has an OpenAPI type, the type must be included in DeclTypes.
+	EnvOptions []cel.EnvOption
+	// ProgramOptions provides CEL ProgramOptions. This may be used to set a cel.CostLimit,
+	// enable optimizations, and set other program level options that should be enabled
+	// for all programs using this environment.
+	ProgramOptions []cel.ProgramOption
+	// DeclTypes provides OpenAPI type declarations to register with the environment.
+	//
+	// If cel.Variables added to EnvOptions refer to a OpenAPI type, the type must be included in
+	// DeclTypes.
+	DeclTypes []*apiservercel.DeclType
+}
+
+// Extend extends the EnvSet with the given VersionedOptions.
+// The returned EnvSet has the same compatibility version as the EnvSet that was extended.
+func (e *EnvSet) Extend(options ...VersionedOptions) (*EnvSet, error) {
+	if len(options) > 0 {
+		validation, storage, err := e.partitionVersionedOptions(e.compatibilityVersion, options)
+		if err != nil {
+			return nil, err
+		}
+		p, err := e.expressionValidation.Extend(validation...)
+		if err != nil {
+			return nil, err
+		}
+		s, err := e.expressionStorage.Extend(storage...)
+		if err != nil {
+			return nil, err
+		}
+		return &EnvSet{compatibilityVersion: e.compatibilityVersion, expressionValidation: p, expressionStorage: s}, nil
+	}
+	return e, nil
+}
+
+func (e *EnvSet) partitionVersionedOptions(compatVer *version.Version, opts []VersionedOptions) ([]cel.EnvOption, []cel.EnvOption, error) {
+	var validatingEnv []cel.EnvOption
+	var storageEnv []cel.EnvOption
+
+	var validatingProg []cel.ProgramOption
+	var storageProg []cel.ProgramOption
+
+	var validatingTypes []*apiservercel.DeclType
+	var storageTypes []*apiservercel.DeclType
+
+	for _, opt := range opts {
+		if compatVer.AtLeast(opt.IntroducedVersion) && (opt.RemovedVersion == nil || compatVer.LessThan(opt.RemovedVersion)) {
+			validatingEnv = append(validatingEnv, opt.EnvOptions...)
+			validatingProg = append(validatingProg, opt.ProgramOptions...)
+			validatingTypes = append(validatingTypes, opt.DeclTypes...)
+		}
+		if opt.RemovedVersion == nil {
+			storageEnv = append(storageEnv, opt.EnvOptions...)
+			storageProg = append(storageProg, opt.ProgramOptions...)
+			storageTypes = append(storageTypes, opt.DeclTypes...)
+		}
+	}
+
+	validatingOpts, err := e.buildOpts(validatingTypes, validatingEnv, validatingProg)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	storageOpts, err := e.buildOpts(storageTypes, storageEnv, storageProg)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return []cel.EnvOption{validatingOpts}, []cel.EnvOption{storageOpts}, nil
+}
+
+func (e *EnvSet) buildOpts(declTypes []*apiservercel.DeclType, envOpts []cel.EnvOption, progOpts []cel.ProgramOption) (cel.EnvOption, error) {
+	if len(declTypes) > 0 {
+		validatingProvider := apiservercel.NewDeclTypeProvider(declTypes...)
+		validatingProviderOpts, err := validatingProvider.EnvOptions(e.expressionValidation.TypeProvider())
```

Comment:
is it correct `buildOpts` is using `e.expressionValidation` directly? `buildOpts` is used to construct `validatingOpts` and `storageOpts`, so I would have expected it to be passed the env to use 

if that's not correct, add a test that would have caught this?

Reply:
This is a bug. Thank you for catching it. I've expanded the tests to cover (in general) and fixed.